src/core/dynamic_batch_scheduler.cc

```diff
@@ -251,169 +246,217 @@ DynamicBatchScheduler::SchedulerThread(
-        }
-        LOG_VERBOSE(1) << "Delaying scheduler thread for " << model_->Name()
-                       << " until " << delay_cnt
-                       << " queued requests, current total = " << queue_.Size();
-      } else if (queue_.Empty()) {
-        wait_microseconds = default_wait_microseconds;
-      } else if (dynamic_batching_enabled_) {
-        // TODO: This method of blocking the scheduler thread till last request
-        // is not scheduled by the rate limiter does not ensure the batched
-        // request is largest possible. There might be new requests queued on
-        // scheduler while rate limiter deferred the execution on model
-        // instance.
-        //
-        // For dynamic batching if there are no available model instance to
-        // allocate or there is already a request pending to be scheduled by the
-        // rate limiter then block the scheduler thread. This is done to allow
-        // dynamic batcher issue a request with largest batchsize.
-        auto ublock_scheduler = [this,
-                                 &pending_request_rate_limiter]() -> bool {
-          return !(
-              (this->rate_limiter_->AvailableInstanceCount(model_) == 0) ||
-              (pending_request_rate_limiter.load()));
-        };
-        cv_.wait(lock, ublock_scheduler);
-        // Use dynamic batching to get request(s) to execute.
-        wait_microseconds = GetDynamicBatch();
-
-        // Get requests that are rejected from searching dynamic batch.
-        queue_.ReleaseRejectedRequests(&rejected_requests);
-
-        // Extract batch only if there is pending batch
-        auto pending_batch_queue_cnt = queue_.PendingBatchCount();
-        if ((wait_microseconds == 0) && (pending_batch_queue_cnt != 0)) {
-          requests.reserve(pending_batch_queue_cnt);
-          for (size_t idx = 0; idx < pending_batch_queue_cnt; ++idx) {
-            std::unique_ptr<InferenceRequest> request;
-            auto status = queue_.Dequeue(&request);
-            if (status.IsOk()) {
-              requests.emplace_back(std::move(request));
-            } else {
-              // The queue is empty which conflicts with pending batch count.
-              // Send the current batch if any and reset related variables.
-              LOG_ERROR << "Failed to retrieve request from scheduler queue: "
+    RateLimiter::ModelInstance* allocated_instance = nullptr;
+    state.store(SchedState::RETRIEVING_REQUEST);
+
+    while (state != READY_FOR_NEW_REQUESTS && !scheduler_thread_exit_.load()) {
+      switch (state.load()) {
+        case SchedState::RETRIEVING_REQUEST: {
+          uint64_t wait_microseconds = 0;
+
+          // Hold the lock for as short a time as possible.
+          {
+            std::unique_lock<std::mutex> lock(mu_);
+            if (delay_cnt > 0) {
+              // Debugging/testing... wait until queue contains 'delay_cnt'
+              // items...
+              wait_microseconds = 10 * 1000;
```

Comment:
Do we want to hardcode this wait?

Reply:
I have just ported this logic from our current dynamic batcher implementation. I have tried to keep the control flow for single instance as similar as possible.